### ISSUE TITLE
For discussion: allow passing a custom function for array instantiation to raster layers

### ISSUE THREAD

**Issue by jni** [napari/napari#7469]
## Motivation

This is a bit of a shower thought so feel free to shoot it down. 😂 I was thinking about [napariboard-proto](https://github.com/jni/napariboard-proto) and how unsatisfying it is that GPU/grad PyTorch arrays need to be [wrapped](https://github.com/jni/napariboard-proto/blob/6b8cd7689126e6133e0e55b995019240a6c74a4f/napariboard.py#L111-L127) in something like a dask array before we can display them. [It's not a priority at PyTorch to support automatic conversion with `.__array__`](https://github.com/pytorch/pytorch/issues/36560), though they might if we add a `force=False` arg to `np.asarray()`. (I tried to start on this a long time ago but got stuck at the NumPy C interface. It's probably worth trying again.)

## Pitch

One way to resolve this is do add an `instantiation_function` kwarg to our raster layers, which defaults to `np.asarray()` but could be user-supplied instead. Then people could do:

```
def grad_gpu_to_numpy(torch_tensor):
    return torch_tensor.detach().cpu().numpy()

layer = viewer.add_image(
        pytorch_gpu_array,
        instantiation_function=grad_gpu_to_numpy,
        )
```

## Alternatives

Probably we should do this either way: add a documentation page showing how to wrap PyTorch (or other) arrays to make sure they display in napari. This includes the minimal API for a "duck array" in napari, which is:

- `.shape`
- `.ndim` (Not 100% whether this is actually needed)
- `.dtype`
- `.__array__`
- `.__getitem__` supperting NumPy-style indexing tuples

**Comment by Czaki** [comment]
Why can users not just provide their own wrapper around the torch array with `__getattr__` for most attributes and overload only critical one?

**Comment by Czaki** [comment]
I prefer wrapping over the new parameter. 

Once we will do decomposition of Layers and have `ScalarFieldData` class to store all information about data, we could allow to subclass it with such changes.
